garmin_fit.cc

```diff
@@ -769,6 +880,478 @@ fit_read()
+                     double speed_ms) {
+  gbfputc(kWriteLocalIdRecord, fout);
+  fit_write_timestamp(timestamp);
+  fit_write_position(position_lat);
+  fit_write_position(position_long);
+  if (distance_m >= 0 && distance_m < 42949672.94) {
+    gbfputuint32(distance_m * 100, fout);
+  } else {
+    gbfputuint32(0xffffffff, fout);
+  }
+  if (altitude != unknown_alt && altitude >= -500 && altitude < 12606.8) {
+    gbfputuint16((altitude + 500) * 5, fout);
+  } else {
+    gbfputuint16(0xffff, fout);
+  }
+  if (speed_ms >= 0 && speed_ms < 65.534) {
+    gbfputuint16(speed_ms * 1000, fout);
+  } else {
+    gbfputuint16(0xffff, fout);
+  }
+}
+
+
+static void
+fit_write_file_header(uint32_t file_size, uint16_t crc)
+{
+  gbfputc(kWriteHeaderCrcLen, fout); // Header+CRC length
+  gbfputc(0x10, fout);               // Protocol version
+  gbfputuint16(0x811, fout);         // Profile version
+  gbfputuint32(file_size, fout);     // Length of data records (little endian)
+  gbfputs(".FIT", fout);             // Signature
+  gbfputuint16(crc, fout);           // CRC
+}
+
+
+static void
+fit_write_header_msgs(gpsbabel::DateTime ctime, QString name)
+{  
+  fit_write_message_def(kWriteLocalIdFileId, kIdFileId, fit_msg_fields_file_id);
+  fit_write_message_def(kWriteLocalIdCourse, kIdCourse, fit_msg_fields_course);
+  fit_write_message_def(kWriteLocalIdLap, kIdLap, fit_msg_fields_lap);
+  fit_write_message_def(kWriteLocalIdEvent, kIdEvent, fit_msg_fields_event);
+  fit_write_message_def(kWriteLocalIdCoursePoint, kIdCoursePoint, fit_msg_fields_course_point);
+  fit_write_message_def(kWriteLocalIdRecord, kIdRecord, fit_msg_fields_record);
+  
+  fit_write_msg_file_id(kFileCourse, 1, 0x3e9, ctime);
+  fit_write_msg_course(name, 0);
+}
+
+
+static void
+fit_write_file_finish()
+{
+  // Update data records size in file header
+  gbsize_t file_size = gbftell(fout);
+  if (file_size < kWriteHeaderCrcLen) {
+    fatal(MYNAME ": File %s truncated\n", fout->name);
+  }
+  gbfseek(fout, 0, SEEK_SET);
+  fit_write_file_header(file_size - kWriteHeaderCrcLen, 0);
+  
+  // Update file header CRC
+  uint16_t crc = 0;
+  gbfseek(fout, 0, SEEK_SET);
+  for (unsigned int i = 0; i < kWriteHeaderLen; i++) {
+    int data = gbfgetc(fout);
+    if (data == EOF) {
+      fatal(MYNAME ": File %s truncated\n", fout->name);
+    }
+    crc = fit_crc16(data, crc);
+  }
+  gbfseek(fout, 0, SEEK_SET);
+  fit_write_file_header(file_size - kWriteHeaderCrcLen, crc);
+  
+  // Write file CRC
+  crc = 0;
+  while (true) {
+    int data = gbfgetc(fout);
+    if (data == EOF) {
+      break;
+    }
+    crc = fit_crc16(data, crc);
+  }
+  gbfputuint16(crc, fout);
+}
+
+static void
+fit_collect_track_hdr(const route_head *rte)
+{
+  course.clear();
+}
+
+static void
+fit_collect_trackpt(const Waypoint* waypointp)
+{
+  course.push_back(FitCourseRecordPoint(*waypointp, false));
+}
+
+static void
+fit_collect_track_tlr(const route_head *rte)
```

Comment:
Should we have a generic version of this in, say, route.cc? Recomputing this seems like something we have in a few formats.

Reply:
ACK. This is pretty similar to what we're doing for Garmin units supported by jeeps. I just don't have the slightest idea about how an API for this should look like. Do you have a function in mind which may be called by formats operating on courses instead of tracks and which returns a course for each track (or calls a callback for each course)?